core/arch/riscv/kernel/boot.c

```diff
@@ -0,0 +1,169 @@
+// SPDX-License-Identifier: BSD-2-Clause
+/*
+ * Copyright 2022 NXP
+ */
+
+#include <assert.h>
+#include <compiler.h>
+#include <config.h>
+#include <console.h>
+#include <keep.h>
+#include <kernel/boot.h>
+#include <kernel/linker.h>
+#include <kernel/misc.h>
+#include <kernel/panic.h>
+#include <kernel/thread.h>
+#include <mm/core_memprot.h>
+#include <mm/core_mmu.h>
+#include <mm/tee_mm.h>
+#include <mm/tee_pager.h>
+#include <riscv.h>
+#include <sbi.h>
+#include <stdio.h>
+#include <trace.h>
+#include <util.h>
+
+#include <platform_config.h>
+
+#define PADDR_INVALID               ULONG_MAX
+
+unsigned long _start_addr;
+unsigned long _boot_args[4];
+
+uint32_t sem_cpu_sync[CFG_TEE_CORE_NB_CORE];
+
+void init_sec_mon(unsigned long nsec_entry __maybe_unused)
+{
+	assert(nsec_entry == PADDR_INVALID);
+	/* Do nothing as we don't have a secure monitor */
+}
+
+#ifdef CFG_RISCV_S_MODE
+static void start_secondary_cores(void)
+{
+	int ret;
+	uint32_t i;
+	size_t pos = get_core_pos();
+
+	for (i = 0; i < CFG_TEE_CORE_NB_CORE; i++) {
+		if (i != pos) {
+#ifdef CFG_RISCV_SBI
+			ret = sbi_boot_hart(i, _start_addr, (unsigned long)i);
+#else
+			ret = -1;
+#endif /*CFG_RISCV_SBI*/
+			if (ret != 0)
+				EMSG("Error starting secondary hart %u", i);
+		}
+	}
+}
+#endif
+
+static void init_runtime(void)
+{
+	malloc_add_pool(__heap1_start, __heap1_end - __heap1_start);
+
+	IMSG_RAW("\n");
+}
+
+void init_tee_runtime(void)
+{
+	core_mmu_init_ta_ram();
+	call_preinitcalls();
+	call_initcalls();
+}
+
+static void init_primary(unsigned long pageable_part __unused,
+			 unsigned long nsec_entry)
+{
+	/*
+	 * Mask asynchronous exceptions before switch to the thread vector
+	 * as the thread handler requires those to be masked while
+	 * executing with the temporary stack. The thread subsystem also
+	 * asserts that the foreign interrupts are blocked when using most of
+	 * its functions.
+	 */
+	thread_set_exceptions(THREAD_EXCP_ALL);
+
+	init_runtime();
+	thread_init_boot_thread();
+	thread_init_primary();
+	thread_init_per_cpu();
+	init_sec_mon(nsec_entry);
+}
+
+/* May be overridden in plat-$(PLATFORM)/main.c */
+__weak void plat_primary_init_early(void)
+{
+}
+
+/* May be overridden in plat-$(PLATFORM)/main.c */
+__weak void main_init_plic(void)
+{
+}
+
+/* May be overridden in plat-$(PLATFORM)/main.c */
+__weak void main_secondary_init_plic(void)
+{
+}
+
+/*
+ * Note: this function is weak just to make it possible to exclude it from
+ * the unpaged area so that it lies in the init area.
+ */
+void __weak boot_init_primary_early(unsigned long pageable_part __unused,
```

Comment:
This function don't have to be `__weak` since riscv doesn't support paging. Same below for `boot_init_primary_late()`

Reply:
Is `__weak ` related to paging only ? my understanding exporting this function as `__weak ` such cab be overridden, by platform/main.c for example.

Reply:
Sure, but in this case it's pretty clear from the comment what was the intent with `__weak` for this function.

Reply:
Comment removed.

Reply:
> Comment removed.

That wasn't my point. You seem to try to keep a bit of the structure from arch/arm booting. In arch/arm platforms are not supposed to override `boot_init_primary_early()`. So why do you insist that it must be possible in arch/riscv?

Reply:
Ok removed.